ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -376,28 +376,45 @@ private[kvutils] class TransactionCommitter(
   private def buildFinalLogEntry: Step =
     (commitContext, transactionEntry) => StepStop(buildLogEntry(transactionEntry, commitContext))
 
-  private def validateContractKeys: Step = (commitContext, transactionEntry) => {
+  private[committer] def validateContractKeys: Step = (commitContext, transactionEntry) => {
     val damlState = commitContext.collectInputs {
       case (key, Some(value)) if key.hasContractKey => key -> value
     } ++ commitContext.getOutputs
-    val startingKeys = damlState.collect {
-      case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty => k
-    }.toSet
-    validateContractKeyUniqueness(commitContext.recordTime, transactionEntry, startingKeys) match {
-      case StepContinue(transactionEntry) =>
-        validateContractKeyCausalMonotonicity(
-          commitContext.recordTime,
-          transactionEntry,
-          startingKeys,
-          damlState)
-      case err => err
+
+    val currentKeyContractIdMapping: Map[DamlStateKey, String] = damlState.collect {
+      case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty =>
+        k -> v.getContractKeyState.getContractId
+    }
+    val startingKeys: Set[DamlStateKey] = currentKeyContractIdMapping.keySet
+    val currentContractKeysMapping: Map[DamlContractKey, String] =
+      currentKeyContractIdMapping.map(m => m._1.getContractKey -> m._2)
+
+    List[() => StepResult[DamlTransactionEntrySummary]](
+      validateContractKeyUniqueness(commitContext.recordTime, transactionEntry, startingKeys),
+      validateContractKeyCausalMonotonicity(
+        commitContext.recordTime,
+        transactionEntry,
+        startingKeys,
+        damlState),
+      validateContractKeysConsistency(
+        commitContext.recordTime,
+        transactionEntry,
+        currentContractKeysMapping),
+    ).foldLeft[StepResult[DamlTransactionEntrySummary]](StepContinue(transactionEntry)) {
+      case (stepResult, check) =>
+        stepResult match {
+          case _: StepContinue[DamlTransactionEntrySummary] => check()
+          case stopResult: StepStop => stopResult
+        }
     }
```

Comment:
I have the suspicion that this `foldLeft` could be written nicely as a `for` comprehension if `StepResult` is given (or already has) a `flatMap` method. But that's maybe for another PR.